File: 10-01-iterator-graph/adts/List.java
package adts;

public interface List<T> extends Iterable<T> {
  public void insert(T data);

  public int getPos(T data);

  public T get(int pos);

  public void remove(T data);

  public void removeAt(int pos);

  public int size();

  public boolean isEmpty();

  public boolean contains(T data);
}

File: 10-01-iterator-graph/adts/list.cpp
#pragma once
#include "./iterator.cpp"

template <class T>
class list : public iterable<T>
{
public:
  virtual void insert(T data) = 0;

  virtual int getPos(T data) = 0;

  virtual T get(int pos) = 0;

  virtual void remove(T data) = 0;

  virtual void removeAt(int pos) = 0;

  virtual bool contains(T data) = 0;

  virtual int size() = 0;

  virtual bool isEmpty() = 0;
};
